src/javascript/app/pages/user/account/settings/two_factor_authentication.js

```diff
@@ -0,0 +1,130 @@
+        $two_factor_loading = $('#two_factor_loading');
+        $qrcode_loading     = $('#qrcode_loading');
+        $qrcode_key         = $('#qrcode_key');
+
+        init();
+    };
+
+    const init = () => {
+        BinarySocket.send({ account_security: 1, totp_action: 'status'}).then((res) => {
+            $two_factor_loading.setVisibility(0);
+
+            if (res.error) {
+                handleError('status', res.error.message);
+                return;
+            }
+
+            current_state = state[res.account_security.totp.is_enabled];
+            next_state    = state[+(!res.account_security.totp.is_enabled)].slice(0, -1);
+
+            $(`#${current_state}`).setVisibility(1);
+            $btn_submit.text(localize(next_state));
+            $form.setVisibility(1);
+
+            FormManager.init(form_id, [
+                { selector: '#otp', validations: ['req', 'number', ['length', { min: 6, max: 6 }]], request_field: 'otp', no_scroll: true, clearResErrorOnInput: true },
+                { request_field: 'account_security', value: 1 },
+                { request_field: 'totp_action',      value: next_state },
+            ]);
+            FormManager.handleSubmit({
+                form_selector       : form_id,
+                fnc_response_handler: handleSubmitResponse,
+                enable_button       : true,
+            });
+
+            if (current_state === 'disabled') {
+                $form.addClass('padding-left-medium');
+                initQRCode();
+            }
+        });
+    };
+
+    const resetComponent = () => {
+        $(`#${current_state}`).setVisibility(0);
+        $form.setVisibility(0).removeClass('padding-left-medium');
+        $qrcode_key.text('');
+
+        $two_factor_loading.setVisibility(1);
+        $qrcode_loading.setVisibility(1);
+
+        init();
+    };
+
+    const initQRCode = () => {
+        BinarySocket.send({ account_security: 1, totp_action: 'generate'}).then((res) => {
+            $qrcode_loading.setVisibility(0);
+
+            if (res.error) {
+                handleError('generate', res.error.message);
+                return;
+            }
+            const secret_key = res.account_security.totp.secret_key;
+            $qrcode_key.text(secret_key);
+
+            makeQrCode(secret_key);
+        });
+    };
+
+    const makeQrCode = (secret_key) => {
+        $('#qrcode').html('');
+        const text = `otpauth://totp/${Client.get('email')}?secret=${secret_key}&issuer=Binary.com`;
+        const qrcode = new QRCode(document.getElementById('qrcode'), {  // eslint-disable-line no-unused-vars
+            text,
+            width       : 160,
+            height      : 160,
+            correctLevel: QRCode.CorrectLevel.H,
+        });
+    };
+
+    const handleSubmitResponse = (res) => {
+        if ('error' in res) {
+            showFormMessage(getPropertyValue(res, ['error', 'message']) || 'Sorry, an error occurred while processing your request.');
```

Comment:
might be worth storing `'Sorry, an error occurred while processing your request.'` in a variable to use here and on line 110